quickwit-metastore/src/metastore/postgresql_metastore.rs

```diff
@@ -0,0 +1,1120 @@
+        })?;
+
+        Ok(())
+    }
+
+    async fn replace_splits<'a>(
+        &self,
+        index_id: &str,
+        new_split_ids: &[&'a str],
+        replaced_split_ids: &[&'a str],
+    ) -> MetastoreResult<()> {
+        let conn = self
+            .connection_pool
+            .get()
+            .map_err(|err| MetastoreError::InternalError {
+                message: "Failed to get connection".to_string(),
+                cause: anyhow::anyhow!(err),
+            })?;
+
+        // Check for the existence of index.
+        let index_exists: bool =
+            self.is_index_exist(&conn, index_id)
+                .map_err(|err| MetastoreError::InternalError {
+                    message: "Failed to publish splits".to_string(),
+                    cause: anyhow::anyhow!(err),
+                })?;
+        if !index_exists {
+            return Err(MetastoreError::IndexDoesNotExist {
+                index_id: index_id.to_string(),
+            });
+        }
+
+        // Check for the inclusion of non-publishable split IDs.
+        // Except for SplitState::Staged and SplitState::Published, you cannot publish.
+        let non_publishable_split_ids = self
+            .get_non_publishable_splits(&conn, index_id, new_split_ids)
+            .map_err(|err| MetastoreError::InternalError {
+                message: "Failed to publish splits".to_string(),
+                cause: anyhow::anyhow!(err),
+            })?;
+        if !non_publishable_split_ids.is_empty() {
+            if let Some(non_publishable_split_id) = non_publishable_split_ids.first() {
+                return Err(MetastoreError::SplitIsNotStaged {
+                    split_id: non_publishable_split_id.clone(),
+                });
+            }
+        }
+
+        let mut error_split_id = "";
+        conn.transaction::<_, diesel::result::Error, _>(|| {
+            // Publish splits.
+            let published_split_ids = self.publish_splits(&conn, index_id, new_split_ids)?;
+
+            // Find out the ID of the update error and return an error if there is one.
+            if published_split_ids.len() < new_split_ids.len() {
+                for split_id in new_split_ids.iter() {
+                    if !published_split_ids
+                        .iter()
+                        .any(|published_split_id| published_split_id == *split_id)
+                    {
+                        error_split_id = split_id;
+                        break;
+                    }
+                }
+                return Err(diesel::result::Error::NotFound);
+            }
+
+            // Mark as deleted.
+            let mark_as_deleted_split_ids =
+                self.mark_as_deleted(&conn, index_id, replaced_split_ids)?;
+
+            // Find out the ID of the update error and return an error if there is one.
+            if mark_as_deleted_split_ids.len() < replaced_split_ids.len() {
+                for split_id in replaced_split_ids.iter() {
```

Comment:
Editing split_error and returning an error, that will be converted afterwards is a very ugly pattern. If you have ideas for refactoring, it might be better.